### PR TITLE
New branch sdk47

### PR DESCRIPTION
**Pull request by sontrinh16**
Closes: #852 

# Description

-remove the third_party folder
-Using sdk47, ibc-go v7
-Switch to forked vesion for ibc-go to bypass the passing of pointer to empty struct in IBCKeeper.NewKeeper
-Switch to forked vesion for sdk47 to bypass valset check when init chain for consumer app, modify expected Slash func parameters in stakingkeeper interface so consumer app can use is consumerkeeper as a stakingkeeper

## Type of change

If you've checked more than one of the first three boxes, consider splitting this PR into multiple PRs!

- [x] Move sdk version to v0.47.1
- [x] Fix all dependencies issues
- [ ] Refactor code base on said changes
- [ ] Remove the legacy-ibc-testing ( base on the TO DO note )
- [ ] Add any migrations logic if needed

## Versioning Implications

- [ ] This PR will affect [semantic versioning as defined for ICS](../CONTRIBUTING.md#semantic-versioning)

If the above box is checked, which version should be bumped?

- [x] `MAJOR`: Consensus breaking changes to both the provider and consumers(s), including updates/breaking changes to IBC communication between provider and consumer(s)
- [ ] `MINOR`: Consensus breaking changes which affect either only the provider or only the consumer(s)
- [ ] `PATCH`: Non consensus breaking changes

## Targeting

Please select one of the following:

- [x] This PR is only relevant to main
- [ ] This PR is relevant to main, and should also be back-ported to ____ (ex: v1.0.0 and v1.1.0)
- [ ] This PR is only relevant to ____ (ex: v1.0.0, v1.1.0, and v1.2.0)

### CODE REVIEW

**Comment by faddat** [comment]
If those are my commits changing the port to 26657, they are wrong. If you look in 812, what you're going to find is sort of like the ideal starting place. 812 is a very clean shift away from SPM. It's possible that the place that you started, did not have that. I think you might be able to try rebasing but in my experience doing a rebase is just a world of pain.

But it might be worth trying.

**Comment by faddat** [comment]
Keep in mind that all the tests in here, you can run them with a verbose flag,

go test ./... --verbose

It's handy for debugging the tests. Also look in the make file, there are additional types of tests and I think that the one that you really want to pass is like

go test integration ./... --parallel --verbose

**Comment by faddat** [comment]
@sontrinh16 -- so to run the tests on your machine you do like:

```bash
 go run ./tests/integration/... --include-multi-consumer --parallel --verbose
```

https://github.com/cosmos/interchain-security/pull/822


I recommend that you continue from 822 -- basically the part of the tests where they call bash, then call jq, needs to be refactored and then it is possible that your branch is finished.  Good work!

**Comment by faddat** [comment]
amazing collaboration @vuong177 @sontrinh16 !!

**Comment by faddat** [comment]
ok, let's try removing the sdk changes as suggested by @mpoke -- also @sontrinh16 can you please mark this as ready for review?  I want to see the test output and the CI is configured so that we can't unless you do.  Thank you.

**Comment by faddat** [comment]
@pysel why the force push?

**Comment by faddat** [comment]
@sontrinh16 I'll do what I usually do and handle lints, if you can handle the validator types.  

I think we can get it to pass tests but @vuong177 says that we will still need to use the sdk fork.

**Comment by sontrinh16** [comment]
yes, consumer app doesnot have a staking module register so there will be no valset in the genesis state which will not pass the init genesis function. So we need to remove the checking process when we init chain for consumer app ( fork the sdk )

**Comment by faddat** [comment]
@vuong177 was correct about the sdk.

Let's resume using his work.

**Comment by faddat** [comment]
@vuong177 we've restarted using the sdk fork you made, thank you very much

**Comment by sontrinh16** [comment]
we still need to have the deprecated keytable as a temporary solution right now for the testcase. Some params still return not registed error

**Comment by pysel** [comment]
> we still need to have the deprecated keytable as a temporary solution right now for the testcase. Some params still return not registed error

Gocha will revert

**Comment by faddat** [comment]
@sontrinh16 can you please edit the description of this PR to describe the cosmos-sdk changes that we needed to make?

**Comment by faddat** [comment]
@smarshall-spitzbart @mpoke @jtremback hey guys, this branch has gotten pretty mature, it passes 287/317 tests.  Could you please look over it?  I am merging main again now, and I also suspect that we could be doing simple things wrong in a few places.

File: x/ccv/consumer/keeper/validators.go [hunk 1]
```diff
@@ -104,23 +102,29 @@ func (k Keeper) ValidatorByConsAddr(sdk.Context, sdk.ConsAddress) stakingtypes.V
...
+) {
+	if infraction == stakingtypes.Infraction_INFRACTION_UNSPECIFIED {
 		return
 	}
 
 	// If this is a previously standalone chain and infraction happened before the changeover was completed,
 	// slash only on the standalone staking keeper.
 	if k.IsPrevStandaloneChain() && infractionHeight < k.FirstConsumerHeight(ctx) {
-		k.standaloneStakingKeeper.Slash(ctx, addr, infractionHeight, power, slashFactor, stakingtypes.InfractionEmpty)
+		k.standaloneStakingKeeper.Slash(ctx, addr, infractionHeight, power, slashFactor)
```
Comment (faddat) [inline x/ccv/consumer/keeper/validators.go:119]: @sontrinh16, @pysel: I think we need to review upstream changes, they may eliminate the need for SlashForked.  Also, I think that we should update godocs here, but I don't know what the difference between SlashForked and Slash is right now.  Thanks!

**Comment by faddat** [comment]
@vuong177 @pysel @sontrinh16 I'd like to focus on fixing this:

```log
    /Users/faddat/interchain-security/tests/integration/common.go:262:
                Error Trace:    /Users/faddat/interchain-security/tests/integration/common.go:262
                                                        /Users/faddat/interchain-security/tests/integration/distribution.go:92
                Error:          Received unexpected error:
                                packet commitment does not exist on either endpoint for provided packet
                Test:           TestCCVTestSuite/TestSendRewardsRetries
                Messages:       error while relaying packets; %!s(MISSING)
```

I think that we aren't sending a packet correctly.  Fixing this would make for many more green tests.

**Comment by pysel** [comment]
> I think that we aren't sending a packet correctly. Fixing this would make for many more green tests.

I will look into this, leaving the comment to avoid parallel work on the same thing

**Comment by faddat** [comment]
eh, I think parallel work could be needed, let's discuss issues at:

https://t.me/+cDuEIPMPBXVkNDU1

**Comment by faddat** [comment]
@pysel great call on removing third party.  Please make sure to rebuild the protobuf deps too.

**Comment by pysel** [comment]
> @pysel great call on removing third party. Please make sure to rebuild the protobuf deps too.

there are some errors with rebuilding it, I do not want to solve it right now, it is not that important at the moment, hence I just reverted it

I say lets revisit it after we deal with the upgrade and start cherry picking

**Comment by faddat** [comment]
I disagree, the protos are god to the system.  I think it is a great idea to make the proto build match ibc-go's if we can:

* https://github.com/cosmos/ibc-go/tree/main/proto

The only trouble is that fork of the sdk that we have here.  Would love to get a reply on the status of that from informal or the sdk team so that we can get moving again.

**Comment by faddat** [comment]
note I need to fix the ics23 stuff..... brb

File: x/ccv/consumer/keeper/validators.go (hunk 1 above)
Comment (sontrinh16) [reply x/ccv/consumer/keeper/validators.go:119]: oh @vuong177 is working on all the slash stuff right now

Comment (sontrinh16) [reply x/ccv/consumer/keeper/validators.go:119]: just need to wait for him to push

**Comment by faddat** [comment]
hey, cool!  so, this is improving I think.

File: .github/workflows/gosec.yml [hunk 2]
```diff
@@ -1,20 +0,0 @@
-name: gosec
```
Comment (jtremback) [inline .github/workflows/gosec.yml:1]: Why is this being deleted?

File: docs/docs/consumer-development/consumer-chain-governance.md [hunk 3]
```diff
@@ -3,23 +3,3 @@ sidebar_position: 2
 ---
 
 # Consumer Chain Governance
-
```
Comment (jtremback) [inline docs/docs/consumer-development/consumer-chain-governance.md:6]: Why is this being deleted?

File: docs/docs/features/key-assignment.md [hunk 4]
```diff
@@ -41,7 +41,7 @@ consumerd tendermint show-validator # {"@type":"/cosmos.crypto.ed25519.PubKey","
 Then, make an `assign-consensus-key` transaction on the provider chain in order to inform the provider chain about the consensus key you will be using for a specific consumer chain.
 
 ```
-gaiad tx provider assign-consensus-key <consumer-chain-id> '<pubkey>' --from <tx-signer> --home <home_dir> --gas 900000 -b block -y -o json
+gaiad tx provider assign-consensus-key <consumer-chain-id> '<pubkey>' --from <tx-signer> --home <home_dir> --gas 900000 -b sync -y -o json
```
Comment (jtremback) [inline docs/docs/features/key-assignment.md:44]: Oh, did they change the flag in 47?

File: go.mod [hunk 5]
```diff
@@ -1,158 +1,183 @@
...
 	github.com/golang/protobuf v1.5.3
 	github.com/gorilla/mux v1.8.0
 	github.com/grpc-ecosystem/grpc-gateway v1.16.0
 	github.com/kylelemons/godebug v1.1.0
 	github.com/oxyno-zeta/gomock-extra-matcher v1.1.0
 	github.com/rakyll/statik v0.1.7
-	github.com/regen-network/cosmos-proto v0.3.1
 	github.com/spf13/cast v1.5.0
-	github.com/spf13/cobra v1.7.0
+	github.com/spf13/cobra v1.6.1
```
Comment (jtremback) [inline go.mod:22]: Why the downgrade?

File: go.mod [hunk 6]
```diff
@@ -1,158 +1,183 @@
...
 )
 
+require github.com/confio/ics23/go v0.9.0 // indirect
+
 replace (
-	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.45.15-ics
-	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
-	github.com/tendermint/tendermint => github.com/cometbft/cometbft v0.34.27
-	google.golang.org/grpc => google.golang.org/grpc v1.33.2
+	github.com/cosmos/cosmos-sdk => github.com/notional-labs/cosmos-sdk v0.47.2-0.20230414131805-781d3665d7a0
```
Comment (jtremback) [inline go.mod:178]: We're going to need to get these into mainline SDK. Can you link your PR on SDK?

File: go.mod [hunk 7]
```diff
@@ -1,158 +1,183 @@
...
+	github.com/cosmos/ibc-go/v7 => github.com/notional-labs/ibc-go/v7 v7.0.0-rc0.0.20230417042817-8072b1e9aabc
```
Comment (jtremback) [inline go.mod:179]: Same. Can you link your PR on IBC?

File: x/ccv/consumer/ibc_module.go [hunk 8]
```diff
@@ -38,7 +39,7 @@ func (am AppModule) OnChanOpenInit(
 
 	// ensure provider channel hasn't already been created
 	if providerChannel, ok := am.keeper.GetProviderChannel(ctx); ok {
-		return "", sdkerrors.Wrapf(types.ErrDuplicateChannel,
+		return "", errorsmod.Wrapf(types.ErrDuplicateChannel,
```
Comment (jtremback) [inline x/ccv/consumer/ibc_module.go:42]: What is this change for?

File: x/ccv/consumer/ibc_module.go [hunk 9]
```diff
@@ -72,41 +73,41 @@ func (am AppModule) OnChanOpenInit(
...
-func (am AppModule) OnChanOpenTry(
-	ctx sdk.Context,
-	order channeltypes.Order,
-	connectionHops []string,
-	portID,
-	channelID string,
-	chanCap *capabilitytypes.Capability,
-	counterparty channeltypes.Counterparty,
-	counterpartyVersion string,
+func (AppModule) OnChanOpenTry(
```
Comment (jtremback) [inline x/ccv/consumer/ibc_module.go:100]: Please put formatting changes in a separate PR. We merged Notional's linting and formatting PRs a couple weeks ago. Why wasn't this changed then?

File: x/ccv/consumer/ibc_module.go [hunk 10]
```diff
@@ -72,41 +73,41 @@ func (am AppModule) OnChanOpenInit(
...
+	_ sdk.Context,
```
Comment (jtremback) [inline x/ccv/consumer/ibc_module.go:101]: Please put formatting changes in a separate PR. We merged Notional's linting and formatting PRs a couple weeks ago. Why wasn't this changed then?

File: x/ccv/consumer/keeper/distribution.go [hunk 11]
```diff
@@ -113,15 +113,17 @@ func (k Keeper) SendRewardsToProvider(ctx sdk.Context) error {
...
-				tstProviderAddr,
-				providerAddr,
-				timeoutHeight,
-				timeoutTimestamp,
-			)
+			packetTransfer := &transfertypes.MsgTransfer{
+				SourcePort:       transfertypes.PortID,     // packet destination port is now the source
+				SourceChannel:    ch,                       // packet destination channel is now the source
+				Token:            token,                    // balance of the coin
+				Sender:           tstProviderAddr.String(), // recipient is the address in the Evmos chain
```
Comment (jtremback) [inline x/ccv/consumer/keeper/distribution.go:120]: These comments look wrong

File: x/ccv/consumer/keeper/distribution.go [hunk 12]
```diff
@@ -211,7 +213,7 @@ func (k Keeper) GetEstimatedNextFeeDistribution(ctx sdk.Context) types.NextFeeDi
 	totalTokens := sdk.NewDecCoinsFromCoins(total...)
 	// truncated decimals are implicitly added to provider
 	consumerTokens, _ := totalTokens.MulDec(frac).TruncateDecimal()
-	providerTokens := total.Sub(consumerTokens)
+	providerTokens := total.Sub(consumerTokens[0])
```
Comment (jtremback) [inline x/ccv/consumer/keeper/distribution.go:216]: What does this do?

File: x/ccv/consumer/keeper/keeper.go [hunk 13]
```diff
@@ -209,11 +212,11 @@ func (k Keeper) SetProviderChannel(ctx sdk.Context, channelID string) {
...
 func (k Keeper) GetProviderChannel(ctx sdk.Context) (string, bool) {
 	store := ctx.KVStore(k.storeKey)
-	channelIdBytes := store.Get(types.ProviderChannelKey())
-	if len(channelIdBytes) == 0 {
+	channelIDBytes := store.Get(types.ProviderChannelKey())
+	if len(channelIDBytes) == 0 {
 		return "", false
 	}
-	return string(channelIdBytes), true
+	return string(channelIDBytes), true
```
Comment (jtremback) [inline x/ccv/consumer/keeper/keeper.go:219]: Why wasn't this changed in the formatting and linting PRs from Notional that we merged a few weeks ago? Shouldn't be in this PR.

File: x/ccv/consumer/keeper/validators_test.go [hunk 14]
```diff
@@ -25,7 +26,7 @@ func TestApplyCCValidatorChanges(t *testing.T) {
 	// utility functions
 	getCCVals := func() (vals []types.CrossChainValidator) {
 		vals = consumerKeeper.GetAllCCValidator(ctx)
-		return
+		return vals
```
Comment (jtremback) [inline x/ccv/consumer/keeper/validators_test.go:29]: No formatting in this PR please

File: docs/docs/features/key-assignment.md (hunk 4 above)
Comment (pysel) [reply docs/docs/features/key-assignment.md:44]: [deprecated](https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/UPGRADING.md#broadcast-mode)

**Comment by faddat** [comment]
Sorry sir it's been closed thank you for being the only informal team member to give actionable feedback. 

We will start from main.

**Comment by mpoke** [comment]
> This pr is valid, it contains the things needed to upgrade this repository as the issue explains. If you are asking for the work to be split across a few prs, which it seems based off the comment then lets document this.
> 
> Contributors are crucial to our ecosystem and closing valid prs doesnt seem like you want people to work on this repo or issue, if this is the case then it should be said that this issue will or should be handled by the team. Notional has updated a few repositories to 0.47 from my knowledge so its worth taking some advice or help from them

We're happy to take advice/help from contributors. In order for us to accept contributions, we need contributors to follow our contributing guidelines (see CONTRIBUTING.md in the root of the repo). The reason for this is, while others may submit these pull requests, our team is ultimately responsible for maintaining the code that gets merged and eventually released, meaning that our team needs as much context as possible on the nature of all of the changes being merged.

As such, for large/complex changes, we absolutely require up-front discussion in issues/GitHub discussions. In some cases, like this change, it would be preferable to even have synchronous design sessions to socialize a collective understanding of the nature of incoming changes prior to the work even starting.

Then, when large changes are submitted, we need those changes to be easy for our team to review. This means that changes either need to be submitted in multiple small PRs (300-500 lines per PR max) that build on top of one another, or that large PRs need to be structured such that they can be reviewed commit by commit (300-500 lines per commit max), with each commit performing a single logical task.

And finally, if someone from our team is assigned to a particular issue, we will most likely opt to not accept unsolicited PRs that attempt to address those same issues. As a general rule, if someone else is assigned to an issue and you would like to take it over, please reach out to that person and ask whether you can either collaborate with them on it or take it over.

We're of course open to collaborating in future, but future collaboration needs to be mindful of the above.

**Comment by faddat** [comment]
Folks, let me make very clear please, and as politely as possible, you don't own this repository, you work on it.

You repeatedly told us that changes to the SDK would not be necessary. We have attempted to get engaged in slack and the like, we have attempted to set up more routine calls with you, we have changed the staffing of this work to attempt to better cater to the needs of informal systems, but the reality is you just keep doing it, and you know, it's really a very perfect scenario for you guys to keep lifting our work over and over and over and over and over again for you to ask us to do the work, have the whole ecosystem. Very excited about getting 47 on ICS, and then simply say that you won't merge it.

Here is where we can have discussions in public:

https://t.me/+cDuEIPMPBXVkNDU1

and of course if you choose not to join well, that should tell the public some stuff, and if you try to say that you're responsive and slack but you're not, I'll publish that as well because the community deserves transparency.

The community should know that you have laid sole claim to this repository.

Now please keep in mind that that governance server is tightly moderated. No one is going to harass you or do nasty stuff there like the previous ones, and if they do they magically disappear.

I am not trying to be nasty here, I am attempting to tell you what is necessary to actually do multi-stakeholder work.
